File: leetcode/valid-palindrome-ii.py
class Solution:
    def validPalindromeRecursive(self, s: str, deleted=False) -> bool:
        # this is correct, but slow and exceeds recursion depth on longstr
        # print(s, deleted)
        if len(s) <= 1:
            return True

        if s[0] == s[-1]:
            return self.validPalindromeRecursive(s[1:-1], deleted)
        elif s[0] != s[-1] and not deleted:
            return self.validPalindromeRecursive(s[1:], True) or self.validPalindromeRecursive(s[:-1], True)
        else:
            return False

    def validPalindrome(self, s: str, deleted=False) -> bool:
        # print(s, deleted)
        i = 0
        j = len(s) - 1

        while i < j:
            if s[i] == s[j]:
                i += 1
                j -= 1
            elif s[i] != s[j] and not deleted:
                return self.validPalindrome(s[i+1:j+1], True) or self.validPalindrome(s[i:j], True)
            else:
                return False
        return True


s = Solution()

assert s.validPalindrome("") == True
assert s.validPalindrome("a") == True
assert s.validPalindrome("aa") == True
assert s.validPalindrome("ab") == True
assert s.validPalindrome("aba") == True
assert s.validPalindrome("abc") == False
assert s.validPalindrome("abca") == True
assert s.validPalindrome("abbadabbad") == True
assert s.validPalindrome("dabbadabba") == True
assert s.validPalindrome("abcddcbaz") == True
assert s.validPalindrome("zabcddcba") == True
assert s.validPalindrome("abczddcbaz") == False

longstr = ""
with open("valid-palindrome-ii-long.txt") as f:
    longstr = f.read()

assert s.validPalindrome(longstr) == True

File: leetcode/valid-palindrome-ii.cpp
#include <string>
#include <assert.h>
using namespace std;

class Solution {
public:
    bool validPalindrome(string s) {
        return valid(s, 0, s.length() - 1, 1);
    }

    bool valid(string s, int l, int r, int skip) {
        while (l < r && s[l] == s[r]){
            l++;
            r--;
        }

        if (l >= r) return true;
        else {
            if (skip != 0){
                return valid(s, l + 1, r, skip - 1) || valid(s, l, r - 1, skip - 1);
            }
            else {
                return false;
            }
        }
    }
};

int main()
{
    Solution s;
    assert(s.validPalindrome("aba") == true);
    assert(s.validPalindrome("abca") == true);
    assert(s.validPalindrome("abc") == false);
    assert(s.validPalindrome("ececabbacec") == true);

    string a = "";
    for (int i=0; i < 100000; i++){
        a += "a";
    }
    assert(s.validPalindrome(a) == true);
    a[10] = 'b';
    assert(s.validPalindrome(a) == true);
    a[11] = 'b';
    assert(s.validPalindrome(a) == false);
}
